Change api spec to return specific xsd and schematron errors

SebastianRossa opened the pull request:
RISDEV-7562

SebastianRossa committed:
Change api spec to return specific xsd and schematron errors

RISDEV-7562

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+      properties:
+        type:
+          type: string
+          enum: [schematron]
```
the value for this currently looks something like this `/errors/ldml-de-not-schematron-valid/failed-assert/SCH-00050-005`

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+      properties:
+        type:
+          type: string
+          enum: [xsd]
```
the value for this currently looks somethings like this `/errors/ldml-de-not-valid/cvc-pattern-valid`

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+          type: string
+          description: Die Fehlermeldung des XSD-Parsers.
+          example: "cvc-complex-type.2.4.a: Invalid content was found starting with element 'author'."
+        line:
```
lineNumber

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+          type: integer
+          description: Zeilennummer des Fehlers im XML.
+          example: 23
+        column:
```
columnNumber

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+          type: string
+          description: Name der Datei, in der der Fehler gefunden wurde.
+          example: regelungstext-1.xml
+        message:
```
details

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+          type: string
+          description: Name der Datei, in der der Fehler gefunden wurde.
+          example: regelungstext-1.xml
+        message:
```
details

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+          type: string
+          enum: [schematron]
+          example: schematron
+        file:
```
this is something we would still need to add

malte-laukoetter commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -278,3 +278,70 @@ components:
+          type: string
+          enum: [xsd]
+          example: xsd
+        file:
```
this is something we would still need to add

malte-laukoetter reviewed (commented):
I added some comments where this differs from the current implementation

SebastianRossa committed:
Change api spec to return specific xsd and schematron errors

RISDEV-7562

malte-laukoetter committed:
Include dokumentName or eli in validation responses

RISDEV-7562

malte-laukoetter committed:
Fix type examples for validation errors

RISDEV-7562

malte-laukoetter committed:
Update example errors

RISDEV-7562

malte-laukoetter committed:
Use eli instead of file name for SchematronValidationError

It is a lot simpler to determine this - and it should provide the same
information

RISDEV-7562

malte-laukoetter committed:
Add additionalProperties to VerkuendungStatusError

RISDEV-7562

VictorDelCampo commented on doc/backend/norms-everkuendung-api-v1.yaml:
```diff
@@ -270,11 +273,80 @@ components:
+            oneOf:
+              - $ref: "#/components/schemas/XsdValidationError"
+              - $ref: "#/components/schemas/SchematronValidationError"
+        extensions:
```
Do we need these extensions? I thought we wanted to have a "fix" structure of what can be returned by us and we just wanted to have the type, title and details right? And we can just add the extra information we would put into these extensions into the detail error

VictorDelCampo reviewed (dismissed):
The mapper `VerkuendungStatusMapper` hasn't changed and I would have expected that this PR would do so. It still has the old implementation from Sebastian where both title and details are being concatenated into single strings.
The schema also doesn't have yet the possible "errors"

malte-laukoetter committed:
Store verkuendung import exceptions as json string

RISDEV-7582

malte-laukoetter committed:
Add archunit test that all classes are sorted into a layer

RISDEV-0000

malte-laukoetter committed:
Add archunit test that config classes are not accessed externally

RISDEV-0000

malte-laukoetter committed:
Allow access to jackson from util layer

It is used for parsing json, and this is something i think the util
layer should be allowed to do

RISDEV-0000

malte-laukoetter committed:
Fix typos

RISDEV-7617

VictorDelCampo commented on backend/src/main/java/de/bund/digitalservice/ris/norms/application/exception/NotLdmlDeXmlFileException.java:
```diff
@@ -12,7 +12,7 @@ public class NotLdmlDeXmlFileException extends RuntimeException implements Norms
   private final String fileName;
 
   public NotLdmlDeXmlFileException(String fileName) {
```
Shouldn't this exception be part of the CreateVerkuendungUseCase?

malte-laukoetter replied:
moved it :)

VictorDelCampo reviewed (commented):
Looks amazing Malte, specially the serializers (both for response schema as well for persisting to db column) and also the command that gets a NormsAppException. Crazy good!

A minor comment about 1 exception and one thought, i was a bit bothered about having to extend from RuntimeException and implement the NormsAppException. Couldn't we just transform the latter from interface to abstract and extend there the RunTimeException so that our custom exceptions just beed to extend NormsAppException?

malte-laukoetter committed:
Move NotLdmlDeXmlFileException to CreateVerkuendungUseCase

RISDEV-7562

malte-laukoetter commented:
> A minor comment about 1 exception and one thought, i was a bit bothered about having to extend from RuntimeException and implement the NormsAppException. Couldn't we just transform the latter from interface to abstract and extend there the RunTimeException so that our custom exceptions just beed to extend NormsAppException?

I was first doing that but we have some exceptions that actually shouldn't be runtime exceptions (most of the once in `ProcessNormendokumentationspaketUseCase`). So I decided against that and to instead both implement the interface and extend the RuntimeException in the other cases.
